src/Sarif/Visitors/PerRunPerTargetSplittingVisitor.cs

```diff
@@ -0,0 +1,71 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.CodeAnalysis.Sarif.Visitors
+{
+    public class PerRunPerTargetSplittingVisitor : SplittingVisitor
+    {
+        private static ArtifactLocation s_emptyArtifactLocation = new ArtifactLocation();
+
+        private Dictionary<string, SarifLog> _targetMap;
+
+        public PerRunPerTargetSplittingVisitor(Func<Result, bool> filteringStrategy = null) : base(filteringStrategy)
+        {
+        }
+
+        public override Run VisitRun(Run node)
+        {
+            _targetMap = new Dictionary<string, SarifLog>();
+            return base.VisitRun(node);
+        }
+
+        public override Result VisitResult(Result node)
+        {
+            if (!FilteringStrategy(node)) { return node; }
+
+            ArtifactLocation artifactLocation = s_emptyArtifactLocation;
+            if (node.Locations[0].PhysicalLocation?.ArtifactLocation != null)
+            {
+                artifactLocation = node.Locations[0].PhysicalLocation?.ArtifactLocation;
+            }
+
+            if (artifactLocation == null)
+            {
+                throw new InvalidOperationException("Result.Locations.PhysicalLocation.ArtifactLocation is null.");
+            }
+
+            if (!_targetMap.TryGetValue(artifactLocation.Uri.ToString(), out SarifLog sarifLog))
+            {
+                sarifLog = _targetMap[artifactLocation.Uri.ToString()] = new SarifLog()
+                {
+                    Runs = new[]
+                    {
+                        new Run
+                        {
+                            Tool = CurrentRun.Tool,
+                            Invocations = CurrentRun.Invocations,
+                            Results = new List<Result>()
+                        },
+                    }
+                };
+                SplitSarifLogs.Add(sarifLog);
+            }
+
+            if (artifactLocation != null && artifactLocation.Index > -1)
+            {
+                int originalIndex = CurrentRun.GetFileIndex(artifactLocation);
```

Comment:
This seems weird - if the ArtifactLocation has an index, you don't need to look it up, right? originalIndex is artifactLocation.Index.

I believe that GetFileIndex has the side-effect of creating an Artifact in the Run if it doesn't already exist, so you also don't need to get sarifLog.Runs[0].Artifacts[..] because GetFileIndex should.

Reply:
This is the same logic in PerRunPerTargetSplittingVisitor, leaving as is for now and we can consider all group strategy optimizations for the future.